include/swift/AST/DiagnosticEngine.h

```diff
@@ -663,6 +694,10 @@ namespace swift {
     /// but rather stored until all transactions complete.
     llvm::StringSet<llvm::BumpPtrAllocator &> TransactionStrings;
 
+    /// Diagnostic producer to handle the logic behind retriving a diagnostic message,
+    /// from YAML files.
+    Optional<LocalizationProducer> localizationProducer;
```

Comment:
@xedin I created another abstraction layer. Now `DiagnosticEngine` can only produce messages... and there's no need to see the `YAMLDiagnosticProvider` object. Do you think it's a better design?

Reply:
I think a better design would be to make `LocalizationProducer` be a super-class of `YAMLLocalizationProducer`, I don't see a need for abstract `YAMLDiagnostic` or `getDiagnostics` method, I think loading could be all done upon construction of `YAMLLocalizationProducer`.

Reply:
I'm sorry I don't quite understand what I should do. Can you explain more what do you mean? also, why would I create another `YAMLLocalizationProducer`?

Reply:
Here is a hierarchy:

- `LocalizationProducer` (has `getMessageOr` and explicit constructor which requires locale/path)
     |_ `YAMLLocalizationProducer` (aka YAML_Diagnostic_Producer we have now)
     |_ `SerializedLocalizationProducer` (that one we'd implement later which doesn't require reading all data from the file at once).

`LocalizationProducer` here is to define an interface which all actual producer would have to implement and users could access.

YAMLLocalizationProducer should upon construction provide data to `LocalizationProducer` (if we don't want to make `getMessageOr` virtual).

Does this make sense?